### PR TITLE
Bug 1075490 - Fix devtools docking.

### PR DESCRIPTION
**Pull request by Gozala**
Depends on platform change attached here (https://bug1075490.bugzilla.mozilla.org/attachment.cgi?id=8546311)

### CODE REVIEW

File: lib/dev/panel.js [hunk 1]
```diff
@@ -227,11 +227,13 @@ createView.define(Panel, (panel, document) => {
   const frame = document.createElement("iframe");
   setAttributes(frame, {
     "sandbox": "allow-scripts",
-    // It would be great if we could allow remote iframes for sandboxing
-    // panel documents in a content process, but for now platform implementation
-    // is buggy on linux so this is disabled.
+    // We end up using chrome iframe with forced message manager
+    // as fixing a swapFrameLoader seemd like a giant task (see
```
Comment (jryans) [inline lib/dev/panel.js]: Nit: seemd -> seemed

File: lib/dev/panel.js [hunk 2]
```diff
@@ -227,11 +227,13 @@ createView.define(Panel, (panel, document) => {
...
+    // Bug 1075490).
     // "remote": true,
-    "type": "content",
+    // "type": "content",
```
Comment (erikvold) [inline lib/dev/panel.js]: nit remove this comment

File: lib/dev/frame-script.js [hunk 3]
```diff
@@ -72,6 +72,9 @@ addMessageListener("sdk/event/message", onInMessage);
 addMessageListener("sdk/port/message", onInPort);
 
 const observer = {
+  handleEvent: ({target, type}) => {
+    this.observe(content.document, type);
```
Comment (erikvold) [inline lib/dev/frame-script.js]: I'm guessing that `this.observe` is undefined here? @gozala does that make sense? and we should have a test that this works.

File: lib/dev/frame-script.js [hunk 4]
```diff
@@ -110,6 +113,6 @@ const observer = {
 
 observerService.addObserver(observer, "content-document-interactive", false);
 observerService.addObserver(observer, "content-document-loaded", false);
-observerService.addObserver(observer, "content-page-hidden", false);
+addEventListener("unload", observer, false);
```
Comment (erikvold) [inline lib/dev/frame-script.js]: where is this defined?

File: lib/dev/panel.js [hunk 5]
```diff
@@ -227,11 +227,13 @@ createView.define(Panel, (panel, document) => {
...
+    "type": "chrome",
```
Comment (erikvold) [inline lib/dev/panel.js]: Do we have a test that remote content is not used?

File: lib/dev/frame-script.js (hunk 4 above)
Comment (Gozala) [inline lib/dev/frame-script.js]: Scope in which frame scripts execute have `addEventListener` defined.

File: lib/dev/panel.js (hunk 5 above)
Comment (Gozala) [inline lib/dev/panel.js]: > Do we have a test that remote content is not used?

I don't think that's relevant to this change, but yes we options are validated:

https://github.com/mozilla/addon-sdk/blob/master/lib/dev/panel.js#L158-L164

File: lib/dev/frame-script.js (hunk 3 above)
Comment (Gozala) [inline lib/dev/frame-script.js]: > I'm guessing that this.observe is undefined here? @gozala does that make sense? and we should have a test that this works.

You are right! I had method before just changed it to arrow to match `observe` & have not run tests again. I'll update that.

Yes we do have tests for unload that were passing until I've changed method to arrow function.

File: test/test-dev-panel.js [hunk 6]
```diff
@@ -150,6 +169,7 @@ exports["test Panel communication"] = test(function*(assert) {
   yield closeToolbox();
 
   assert.equal(panel.readyState, "destroyed", "panel is destroyed");
+  myTool.destroy();
```
Comment (erikvold) [inline test/test-dev-panel.js]: this seems unrelated, if it is then please make a separate bug for this change.

File: test/test-dev-panel.js [hunk 7]
```diff
@@ -86,6 +86,25 @@ exports["test Panel API"] = test(function*(assert) {
   yield closeToolbox();
 
   assert.equal(panel.readyState, "destroyed", "panel is destroyed");
+
+  myTool.destroy();
+});
+
+
```
Comment (erikvold) [inline test/test-dev-panel.js]: we should add a comment mentioning that this is important since we use a "chrome" frame, and include a link to this bug too.

File: test/test-dev-panel.js [hunk 8]
```diff
@@ -86,6 +86,25 @@ exports["test Panel API"] = test(function*(assert) {
...
+exports["test forbid remote docs"] = test(function*(assert) {
+  const MyPanel = Class({
+    extends: Panel,
+    label: "test panel",
+    tooltip: "my test panel",
+    icon: iconURI,
+    url: "https://mozilla.org",
+  });
+
```
Comment (erikvold) [inline test/test-dev-panel.js]: another test for `http` too please.

Comment (erikvold) [inline test/test-dev-panel.js]: another test for `ftp` too please.

**Comment by erikvold** [comment]
We should have a test that it's not possible to have the local content (which passes validation) redirect to non-local content.

**Comment by erikvold** [comment]
This is failing on travis:

7114 of 7118 tests passed.

JPM info There were test failures...
JPM info 
The following tests failed:

JPM info   ./test/test-dev-panel.test Panel API: timed out
  ./test/test-dev-panel.test Panel communication: timed out

JPM info   ./test/test-dev-panel.test communication with debuggee: timed out
  ./test/test-dev-panel.test viewFor panel: timed out

**Comment by erikvold** [comment]
>> I'm guessing that this.observe is undefined here? @gozala does that make sense? and we should have a test that this works.

> You are right! I had method before just changed it to arrow to match `observe` & have not run tests again. I'll update that.

> Yes we do have tests for unload that were passing until I've changed method to arrow function. 

Which ones?

File: lib/dev/frame-script.js [hunk 9]
```diff
@@ -82,21 +85,21 @@ const observer = {
       observerService.removeObserver(observer, topic);
     }
     else if (document === content.document) {
-      if (topic === "content-document-interactive") {
+      if (topic.endsWith("-document-interactive")) {
```
Comment (erikvold) [inline lib/dev/frame-script.js]: this is an unrelated change, please separate this change in to another bug.

File: lib/dev/frame-script.js [hunk 10]
```diff
@@ -110,6 +113,8 @@ const observer = {
 
 observerService.addObserver(observer, "content-document-interactive", false);
 observerService.addObserver(observer, "content-document-loaded", false);
-observerService.addObserver(observer, "content-page-hidden", false);
+observerService.addObserver(observer, "chrome-document-interactive", false);
+observerService.addObserver(observer, "chrome-document-loaded", false);
```
Comment (erikvold) [inline lib/dev/frame-script.js]: why do you need both chrome and content observers here?

File: lib/dev/frame-script.js [hunk 11]
```diff
@@ -72,6 +72,9 @@ addMessageListener("sdk/event/message", onInMessage);
 addMessageListener("sdk/port/message", onInPort);
 
 const observer = {
+  handleEvent: ({target, type}) => {
+    observer.observe(content.document, type);
```
Comment (erikvold) [inline lib/dev/frame-script.js]: can you use the `target` here please, or remove the argument variable.

File: lib/dev/frame-script.js (hunk 9 above)
Comment (Gozala) [inline lib/dev/frame-script.js]: > this is an unrelated change, please separate this change in to another bug.

It is related because if content is chrome privileged topic will be `chrome-document-interactive"

File: lib/dev/frame-script.js (hunk 10 above)
Comment (Gozala) [inline lib/dev/frame-script.js]: > @gozala why do you need both chrome and content observers here?

Because it could be chrome privileged document.

File: test/test-dev-panel.js (hunk 7 above)
Comment (Gozala) [inline test/test-dev-panel.js]: > we should add a comment mentioning that this is important since we use a "chrome" frame, and include a link to this bug too.

I do not understand this. You want me to comment why I destroy a tool here ?

**Comment by Gozala** [comment]
> This is failing on travis:
> 
> 7114 of 7118 tests passed.
> 
> JPM info There were test failures...
> JPM info
> The following tests failed:
> 
> JPM info ./test/test-dev-panel.test Panel API: timed out
> ./test/test-dev-panel.test Panel communication: timed out
> 
> JPM info ./test/test-dev-panel.test communication with debuggee: timed out
> ./test/test-dev-panel.test viewFor panel: timed out

I've already pointed out that travis will fail .as it depend on platform change

**Comment by Gozala** [comment]
> We should have a test that it's not possible to have the local content (which passes validation) redirect to non-local content.

I'm afraid I can't make such guarantees, since we make this local content chrome privileged it can do whatever it pleased no matter how hard we try.

**Comment by Gozala** [comment]
@erikvold I removed those `myTool.destroy()` calls now as well. But honestly I think you are not consistent with your requests, in some cases you ask me to add unrelated tests and in other cases you ask me to remove changes that ensure proper cleanup after test run.

Also now that related platform changes are in nightly travis is happy and passing tests!

**Comment by Gozala** [comment]
Created separate pull for test cleanups as requested here #1847

**Comment by Gozala** [comment]
Also here is a link to unload tests that came up in the discussion #1848

File: lib/dev/frame-script.js [hunk 12]
```diff
@@ -110,6 +113,8 @@ const observer = {
 
 observerService.addObserver(observer, "content-document-interactive", false);
 observerService.addObserver(observer, "content-document-loaded", false);
-observerService.addObserver(observer, "content-page-hidden", false);
+observerService.addObserver(observer, "chrome-document-interactive", false);
+observerService.addObserver(observer, "chrome-document-loaded", false);
+addEventListener("unload", observer, false);
```
Comment (ochameau) [inline lib/dev/frame-script.js]: Shouldn't you remove this listener at some point? Same question for observerService?
Or is this frame script global and want to dispatch events for all document living in the tab?

File: lib/dev/frame-script.js (hunk 10 above)
Comment (ochameau) [inline lib/dev/frame-script.js]: It looks like test-dev-panel only has panels using data: URI.
It would be great if you could also test a chrome:// URL, like `chrome://global/content/about.xhtml` for ex, can we load any document or does it has to do something special to work in panel?
